An app to detect licenses from the provided input license text

lf32 opened the pull request:
The aim of this PR is to create a working app inside this Django project to detect licenses from the provided input text and summarize the results for the end user.

lf32 committed:
create an app to detect licenses from input text

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented on scancodeio/settings.py:
```diff
@@ -45,7 +45,7 @@
 
 # SECURITY WARNING: don't run with debug turned on in production
-DEBUG = env.bool("SCANCODEIO_DEBUG", default=False)
+DEBUG = env.bool("SCANCODEIO_DEBUG", default=True)
```
This setting is not meant to be changed in that file, put `SCANCODEIO_DEBUG=True` in your env variables instead.

lf32 replied:
Yes, will restore it in the upcomming commit

tdruez commented on scancodeio/urls.py:
```diff
@@ -52,6 +52,7 @@
     path("admin/", admin.site.urls),
     path("api/", include(api_router.urls)),
     path("license/", include(licenses.urls)),
+    path("scan/", include("scantext.urls")),
```
`scan` is a bit to generic, let's use `scantext` in the url as well.

lf32 replied:
Sure, will change it to `scantext` in the upcoming commit

tdruez commented on scantext/admin.py:
```diff
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: Apache-2.0
```
This file is not needed.

lf32 replied:
This file will be discarded int the upcoming commit

tdruez commented on scantext/apps.py:
```diff
@@ -0,0 +1,28 @@
+
+
+class ScantextConfig(AppConfig):
+    default_auto_field = "django.db.models.BigAutoField"
```
Are you sure this is needed?

lf32 replied:
Nope ig, it was generated by default

tdruez commented on scantext/templates/scantext/includes/license_summary_level.html:
```diff
@@ -0,0 +1,35 @@
+      <p class="title">
+          <span>
+            {{ expr.license_expressions | length }}
+
```
Empty line

tdruez commented on scantext/templates/scantext/includes/license_summary_level.html:
```diff
@@ -0,0 +1,35 @@
+      <p class="title">
+          <span>
+            {{ expr.licenses | length }}
+
```
Empty line

tdruez commented on scantext/templates/scantext/includes/license_summary_level.html:
```diff
@@ -0,0 +1,35 @@
+      <p class="heading">License Expressions</p>
+      <p class="title">
+          <span>
+            {{ expr.license_expressions | length }}
```
Spaces around "|" are not consistent with others existing templates.

lf32 replied:
yes, will correct them in the upcoming commit

tdruez commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -0,0 +1,107 @@
+          <p class="title is-4">Input License Text</p>
+           <div class="box">
+             {% for line in text %}
+              <p>{{line}}</p>
```
Leave a space after `{{` and before `}}`

tdruez reviewed (commented):
@lf32 That's a good start, see my comments for minor improvements.

In the future you need to provide more context than simply requesting a review.
There's a tons of commented code, which part do you need to be reviewed?
Include what you have accomplished and what is left to do, also include the issues you are facing.

lf32 commented:
> @lf32 That's a good start, see my comments for minor improvements.
> 
> In the future you need to provide more context than simply requesting a review. There's a tons of commented code, which part do you need to be reviewed? Include what you have accomplished and what is left to do, also include the issues you are facing.

thanks, i pinged you to let you know about the PR

- the `scancode.api.get_licenses` accepts files as input but not text, what do you think of it?
- do you think the input text should be stored as a file and then pass it as the argument?

tdruez commented:
> the scancode.api.get_licenses accepts files as input but not text, what do you think of it?

The copied `get_licenses` is not meant to be used as-is to to be modified to fir your need.

> do you think the input text should be stored as a file and then pass it as the argument?

No, update the code to work on text.

tdruez commented on scanpipe/templates/scanpipe/base.html:
```diff
@@ -5,6 +5,7 @@
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
     <title>{% block title %}ScanCode.io{% endblock %}</title>
+    <link rel="icon" href="{% static 'favicon.ico' %}" type="image/x-icon"/>
```
This is unrelated to this PR.

tdruez replied:
You can merge main into your branch once in a while to get the latest changes such as this one that is now in main ;)

lf32 replied:
Thanks;) 
Got a merge conflict :)

tdruez replied:
This conflict should be trivial to fix.

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

Conflicts:
	scanpipe/templates/scanpipe/base.html

lf32 committed:
Merge branch 'main' of nexB/scancode.io into lf32-licensetext-devel

lf32 commented:
@tdruez please suggest changes and also any resource for recommending license?

![res_license](https://user-images.githubusercontent.com/96695352/174466914-40c79b2e-c709-48e8-b107-3214f27689cf.png)

lf32 commented:
> <img alt="Screenshot 2022-06-16 at 12 15 58" width="322" src="https://user-images.githubusercontent.com/489057/174025195-4025be49-fe39-49f7-96f9-e1439a4efaec.png">
> 
> You are putting side by side a form submit button "Scan License" and a navigation link "New Project".
> 
> I think the "New Project" as no need to be present in that view. Also, the action button "Scan License" would make more sense to be located after the form, not on top of it. The textarea is simply too big, reduce the height.

![res_license](https://user-images.githubusercontent.com/96695352/174467562-d1e27804-5fff-4081-9ffe-4e2881231492.png)

How is this?
rows are set to 25, isn't it ok?

tdruez commented:
> please suggest changes

This looks awesome but where is the code? No code, no review possible.

>  any resource for recommending license?

I don't know what you mean by that.

lf32 committed:
Improved UI, Changed Scan Path
* Dropdown cards work well in the details page
* Re-arranged html code
* Changed Scan path from /scan/ to /scantext/

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
> This looks awesome but where is the code? No code, no review possible.

I have just pushed changes over here

> > any resource for recommending license?
> 
> I don't know what you mean by that.
I mean what to recommend if the deteced expression is `unknown-license-reference`

lf32 committed:
changed card title from name to short_name

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez reviewed (changes requested):
@lf32 that's a great start, see my comments for minor improvements.

I think the UI is fine for now, you should focus on the views code to make it working on real input text.

lf32 commented on scanpipe/templates/scanpipe/base.html:
```diff
@@ -57,4 +57,4 @@
     <script src="{% static 'htmx-1.7.0.min.js' %}" crossorigin="anonymous" defer></script>
     {% block scripts %}{% endblock %}
   </body>
-</html>
```
Oh, I haven't noticed it

sure, will fix this

lf32 commented on scantext/forms.py:
```diff
@@ -0,0 +1,42 @@
+    input_text = forms.CharField(
+        widget=forms.Textarea(
+            attrs={
+                "rows": 25,
```
that's great, will set them to 15

lf32 commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -0,0 +1,99 @@
+            <p class="title is-4">Detected Licenses</p>
+            <div class="card">
+                {% for license in expr.licenses %}
+                <div class="licenses-card">
```
how about `license_detail_card`?

tdruez replied:
Sounds good.

lf32 commented on scantext/templates/scantext/license_scan.html:
```diff
@@ -0,0 +1,50 @@
+{% extends "scanpipe/base.html" %} 
```
ok, will change it

lf32 commented on scantext/forms.py:
```diff
@@ -0,0 +1,42 @@
+from django import forms
+
+
+class EditorForm(forms.Form):
```
how about `LicenseScanForm`

lf32 replied:
How about `LicenseScanForm`?

tdruez commented:
@lf32 please read the following for future commit messages (no need to update the existing ones): https://aboutcode.readthedocs.io/en/latest/contributing/writing_good_commit_messages.html

For example: "changed card title from name to short_name" should be "Change card title from name to short_name #450"

You can look at the commit history of the repo and try to follow the conventions in place.

tdruez commented on scantext/views.py:
```diff
@@ -0,0 +1,227 @@
+        form = EditorForm(request.POST)
+        if form.is_valid():
+            text = form.cleaned_data["input_text"]
+            # license_location = tempfile.NamedTemporaryFile(mode="w", prefix="license_scan_", dir=settings.SCANCODEIO_WORKSPACE_LOCATION)
+            # with license_location as f:
+            #     f.write(text)
+            #     f.flush()
+            #     x=get_licenses(location=f)
+            #     f.close()
+            # the get_licenses in the above code (line 56) returns this error
+            # error:
+            # expected str, bytes or os.PathLike object, not _TemporaryFileWrapper
+            # the below code just works
+
+            expressions = get_licenses(
+                location="scantext/tests/data/LICENSES",
+                include_text=True,
+                license_text_diagnostics=True,
+            )
```
What about:
```
            with tempfile.NamedTemporaryFile(mode="w") as temp_file:
                temp_file.write(text)
                expressions = get_licenses(location=temp_file.name)
```

tdruez commented on scantext/templates/scantext/license_scan.html:
```diff
@@ -0,0 +1,50 @@
+                <h1 class="title is-5">Scan License</h1>
+            </div>
+            <div>
+                <a href="{% url 'project_list' %}" class="button is-link">Go to Projects</a>
```
I don't think we need this link anymore since we have a "Projects" link in the header.
Also, we should add a new "Scan" link in the header right next to "Projects" and before "Documentation" that would redirect to the `license_scan` view.
This may change in the future but this will save us some navigation time during the development phase.

lf32 commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -0,0 +1,99 @@
+    </section>
+
+    <hr class="mx-5">
+    {% include "scantext/includes/license_summary_level.html" with expr=expr %}
```
what about `matches`, `summary`, `results`, `details`?

tdruez replied:
`results` is too generic, you want to name your variable so a reader know what it actually contains.
In this case, what about something like `detected_licenses`?

lf32 replied:
`detected_licenses` sounds good

lf32 commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -0,0 +1,99 @@
+                <h1 class="title is-4">License Detection Summary</h1>
+            </div>
+            <div>
+                <a href="{% url 'license_scan' %}" class="button is-link">Scan Again</a>
```
Do you think we need to use this when we have `Scan` option in the header itself

lf32 replied:
sure, will use update this in the upcoming commit

lf32 committed:
Improved License Input and Details UI #450

* Set Input Textarea to 15
* Added file-upload into form
* Changed <pre> to <textarea> in the details page
* Modified Navigation Bar and Renamed Scan Again Button

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Run License Detection On Text Submission with tempfile #450

* license_scanview function uses tempfile to run license detection on the
  provided input license text

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented on scantext/forms.py:
```diff
@@ -0,0 +1,42 @@
+from django import forms
+
+
+class LicenseForm(forms.Form):
```
`LicenseScanForm` as you suggested was a better name.

tdruez commented on scantext/forms.py:
```diff
@@ -0,0 +1,42 @@
+
+
+class LicenseForm(forms.Form):
+    input_text = forms.CharField(
```
We may want to disable the default `strip` on this field because we are loosing some of the initial format when displaying the text in the results view.
See the first line for example:
![Screenshot 2022-06-21 at 11 07 25](https://user-images.githubusercontent.com/489057/174737653-d4671e44-b0bb-43a3-b5ee-ded34d9179af.jpg)

What's you take on this?

lf32 replied:
We can set it to false in the form but the get_license does strip the matched_text, which means the card which shows detected text does look the same way above

tdruez reviewed (changes requested):
@lf32 Good progress!

Some feedback:

- Commit messages look better but there's no need to uppercase every single words.
A sentence starts with an upper case then each words is lower case.
Please have a look at the commit history of the repo and try to be consistent with the existing style.

- Also, what's the deal with all the `force-pushed` in your commit history?
This is a source of issues when fetching your changes on my side. Try to avoid those rewriting in the commits history.

- Remove the `models.py` file and `migrations/` directory. Not needed for now.

- It would be nice to highlight the "matched text" on the left side when I open the match details on the right side. Does that feature make sense to you?

lf32 commented on scantext/views.py:
```diff
@@ -0,0 +1,224 @@
+                dir=settings.SCANCODEIO_WORKSPACE_LOCATION,
+            ) as temp_file:
+                temp_file.write(text)
+                temp_file.flush()
+                expressions = get_licenses(
+                    location=temp_file.name,
+                    include_text=True,
+                    license_text_diagnostics=True,
+                )
+                temp_file.close()
```
Oh, I thought I had to use flush() to write the file
Ok, will clean up lines 52 and 58

lf32 commented:
should the user have a checkbox to decide if they have to detect `unknown licenses` or should we just set it to true by default

for the above license the percentage is 99.21 when its true, but 97.91 when set to False

tdruez commented:
> should the user have a checkbox to decide if they have to detect unknown licenses or should we just set it to true by default

It's not yet the time to make those kind of advance refinements.
We want to display everything for now.

lf32 commented on scantext/views.py:
```diff
@@ -0,0 +1,224 @@
+            text = form.cleaned_data["input_text"]
+            with tempfile.NamedTemporaryFile(
+                mode="w",
+                prefix="license_scan_",
+                dir=settings.SCANCODEIO_WORKSPACE_LOCATION,
```
I thought executing it in the `var` would be better
what do you think?
should I remove it?

lf32 replied:
After thinking a bit more, I think you are right

lf32 replied:
I added prefix so that it would be better to understand

tdruez replied:
That file is instantly closed and deleted when exiting the context manager, I still don't see the need for a prefix.
As explained earlier the `close()` call is not needed.

lf32 committed:
Rename license form, prevent text stripping, temp_file attr update #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Detect unknown licenses and return matched text by default #450
* get_licenses detects unknown license by default
* view returns matched_text by default
* change variable attribute from `result` to `detected_licenses`
Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
oops!, had a missing DCO, sorry for that :

lf32 committed:
Flush contents of the tempfile before passing as an argument #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Flush contents of the temfile before passing as an argument #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
> @lf32 your recent changes in `get_licenses` breaks the detection of simple content such as "Apache 2.0". Is this behavior expected?

Yes I guess, when tempfile is not flushed with temp_file.flush() I think the file doesn't get written to disk and get_licenses cant read it from the beginning

lf32 committed:
Merge branch 'lf32-licensetext-devel' of https://github.com/nexB/scancode.io into lf32-licensetext-devel

tdruez commented:
> Yes I guess, when tempfile is not flushed with temp_file.flush() I think the file doesn't get written to disk and get_licenses cant read it from the beginning

Keep the .flush() and add a comment saying that the flush is require to ensure that the content is written to the disk before it's read by get_licenses

tdruez commented:
Also, you haven't answer to my questions from my previous comment:

- Also, what's the deal with all the force-pushed in your commit history?
  This is a source of issues when fetching your changes on my side. Try to avoid those rewriting in the commits history.

- Remove the models.py file and migrations/ directory. Not needed for now.

- It would be nice to highlight the "matched text" on the left side when I open the match details on the right side. Does that feature make sense to you?

lf32 commented:
> @lf32 Good progress!
> 
> Some feedback:

Thanks, for the feedback :)
> 
>     * Commit messages look better but there's no need to uppercase every single words.
>       A sentence starts with an upper case then each words is lower case.
>       Please have a look at the commit history of the repo and try to be consistent with the existing style.
> 
>     * Also, what's the deal with all the `force-pushed` in your commit history?
>       This is a source of issues when fetching your changes on my side. Try to avoid those rewriting in the commits history.
> 

I have force pushed in the past because I was trying to make commits look better and, in a few other cases, I had to amend simple changes to the last commit which was pushed to github.

>     * Remove the `models.py` file and `migrations/` directory. Not needed for now.
> 

lost the commit during a hard reset

>     * It would be nice to highlight the "matched text" on the left side when I open the match details on the right side. Does that feature make sense to you?

Is it like how demo.grammarly.com does?
It sounds great and I'm working on it

I am having issues with accessing

lf32 committed:
Deleted models.py and migrations #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Added comment related to flushing tempfile #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented:
> Is it like how demo.grammarly.com does?
> It sounds great and I'm working on it

That UI looks pretty good, I like that we could use different colors based on the criticality of the various matches.

lf32 committed:
Commenting inside views with pound over docstring #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Write comments in multiple lines inside views #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented on scantext/views.py:
```diff
@@ -43,22 +43,23 @@ def license_scanview(request):
             text = form.cleaned_data["input_text"]
+            """
+            The flush in tempfile is required to ensure that the content is written to the disk before it's read by get_licenses function
+            """
```
sure, I will write comments in multiple lines without crossing line length limit

lf32 commented on scantext/views.py:
```diff
@@ -146,7 +147,7 @@ def _licenses_data_from_match(
     """
-    Returned matched_text will also include the text detected 
+    Returned matched_text will also include the text detected
     """
```
some times `make valid` returns error, any idea on how to fix this?

It says i don't have `bin/isort`

tdruez replied:
Did you run `make dev` first?

lf32 committed:
Input text is rendered inside the box over the textarea #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
> > Is it like how demo.grammarly.com does?
> > It sounds great and I'm working on it
> 
> That UI looks pretty good, I like that we could use different colors based on the criticality of the various matches.

based on score?

tdruez commented:
> based on score?

I was thinking more about [License Policies and Compliance Alerts](https://scancodeio.readthedocs.io/en/latest/tutorial_license_policies.html#license-policies-and-compliance-alerts)

lf32 committed:
Merge branch 'main' of https://github.com/nexB/scancode.io into lf32-licensetext-devel

lf32 committed:
Display form errors, detect licenses for input text file #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -20,13 +20,13 @@ <h1 class="title is-4">License Detection Summary</h1>
         <div class="column is-half">
             <p class="title is-5">Input License Text</p>
-	    <div class="box" style="height: 70vh; overflow-y: scroll; white-space: pre-wrap;">{% for line in text %}<span>{{ line }}</span>{% endfor %}</div>
+	       <div class="box" style="max-height: 68vh; overflow-y: scroll; white-space: pre-wrap;">{% for line in text %}<span >{{ line }}</span>{% endfor %}</div>
```
Un-wanted whitespaces.

tdruez commented on scantext/templates/scantext/license_detail.html:
```diff
@@ -20,13 +20,13 @@ <h1 class="title is-4">License Detection Summary</h1>
             <div class="card">
-                {% include 'scantext/includes/license_detail_card.html' with detected_licenses=detected_licenses %}
-            </div>
+                {% include 'scantext/includes/license_detail_card.html' with detected_licenses=detected_licenses %}     
+            </div>       
```
Un-wanted whitespaces.

tdruez commented on scantext/views.py:
```diff
@@ -40,25 +40,64 @@
-                expressions = get_licenses(
-                    location=temp_file.name,
+            if len(input_text):
+                with tempfile.NamedTemporaryFile(
+                    mode="w",
+                ) as temp_file:
```
This should be on 1 line to make it more readable.

lf32 replied:
sure, will make it readable

tdruez commented on scantext/views.py:
```diff
@@ -40,25 +40,64 @@
+                ) as temp_file:
+                    temp_file.write(input_text)
+                    temp_file.flush()
+                    expressions = get_licenses(
+                        location=temp_file.name,
+                    )
```
Breaking short line does not help readability. This would be better as a one-liner.

lf32 replied:
sure, will make it one-liner

tdruez reviewed (changes requested):
The code works fine. See my comment for improvements.
You should now focus on the UI to help reviewing the results.

lf32 commented:
@tdruez what do you think about [github's changelog?](https://github.blog/changelog/)

tdruez commented:
> @tdruez what do you think about [github's changelog?](https://github.blog/changelog/)

@lf32 In which context? 
I'm not familiar with it.

lf32 commented:
displaying detected license in the way `changelog` looks, instead of highlighting the text

tdruez commented:
@lf32 please explain why you think this UI would be better.

lf32 commented:
In the previous case user has to click on the card's drop down
Here user has to just scroll

lf32 commented:
![res_license](https://user-images.githubusercontent.com/96695352/176398377-91313339-af19-49a3-a269-f832614f30fc.png)

tdruez commented:
> In the previous case user has to click on the card's drop down

Please be explicit with what you mean by "previous case", it's not clear.

> Besides that, there is an issue in detecting matched text from the html because the matched result is actually altered from the original one

This is an issue and it can probably be solved. Changing the end results because your are facing a difficulty is generally not a good enough reason.

Actually, it would be nice to have both options in separate tabs:
- The list and matched text like presented in your screenshot at https://github.com/nexB/scancode.io/pull/450#issuecomment-1169724197 as "Details"
- The full input text with all the matchs highlighted to view all the matches in the context of the whole text.

lf32 commented:
> Please be explicit with what you mean by "previous case", it's not clear.

sure!
> This is an issue and it can probably be solved. Changing the end results because your are facing a difficulty is generally not a good enough reason.

:sweat_smile: Would you please help me with this? I have some issues.
> The list and matched text like presented in your screenshot at ... as "Details"

can we discuss this, I have few questions (on gitter)

tdruez commented:
> 😅 Would you please help me with this? I have some issues.

Sure, can you try to explain the problem, and present some example?

> can we discuss this, I have few questions (on gitter)

Why not here instead of gitter?

lf32 committed:
Improve file error response in views #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented on scantext/views.py:
```diff
@@ -40,25 +40,64 @@
+            input_text = form.cleaned_data["input_text"]
+            input_file = request.FILES.get("input_file", False)
+            if not len(input_text) and not input_file:
+                message = "Please provide some text or a text file to scan."
```
sure, I have used them to return messages

lf32 committed:
Moved license summary to details with  new UI #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Fixed broken short lines #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented on scantext/templates/scantext/includes/license_detail.html:
```diff
@@ -0,0 +1,48 @@
+    }
+
+</style>
+{% endfor %}
```
will fix this in the next commit

lf32 replied:
fixed  it.

lf32 commented on scantext/templates/scantext/license_summary.html:
```diff
@@ -0,0 +1,76 @@
+{% endblock %}
+
+{% block scripts %}
+<script src="{% static 'billboard-3.0.1.pkgd.min.js' %}" crossorigin="anonymous"></script>
```
was trying to present a bar chart in the summary page

lf32 committed:
New line at the end of the page #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented on scantext/views.py:
```diff
@@ -40,25 +40,64 @@
+
+            if not len(expressions["licenses"]) and not len(
+                expressions["license_expressions"]
+            ):
```
Updated the code.

lf32 committed:
Merge branch 'main' of nexB/scancode.io -> lf32-licensetext-devel

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented:
@lf32 could you provide a quick status of your progress?
What is now complete, what you are currently working on, and what is left todo.

Cheers!

lf32 commented:
Progress Update
- Developed Views
- Tweaking around Templates
- Have to discuss on what's next to do.

lf32 commented:
@tdruez I am thinking of using ace editor instead of manually designing the input text div (left column).

Which one do you think is having a better view?
---
![1](https://user-images.githubusercontent.com/96695352/177151163-d5a99c88-45c5-4830-b2ee-ca0ee6079035.png)
---
![2](https://user-images.githubusercontent.com/96695352/177151388-72a0fa92-d6bb-4aad-8aa7-9733f1d72d3a.png)

lf32 committed:
Ace Editor Restored with improved UI #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented:
@lf32 maybe the second one.
But I do not like the "View" and "View Input License" system. We should not replace the whole text by matched text, we already have the list of matched text at the bottom of the page.
Instead we should highlight the "match text" in the context of the whole text.

lf32 committed:
Highlight and scroll down to matched text #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Close all cards while opening a new one #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Removed details option in the view #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Restore details tab with matched text #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

tdruez commented:
@lf32 we haven't heard or seen code from you in over 15 days.
Don't forgot to update us here with weekly reports on your progress and encountered issues.

lf32 commented:
@tdruez yes (sorry for no update), I was going through the ace docs.

The highlighting issue is the same in the resources page of any project. Highlighting happens on matched text though there is some missing part in the matched text.

* Tweaking the views `match.matched_text(whole_lines=False, highlight=False)`
* Going through the ace docs
* Made a PR at #469

lf32 committed:
Merge remote-tracking branch 'origin' into lf32-licensetext-devel

lf32 committed:
Merge remote-tracking branch 'origin' into lf32-licensetext-devel

lf32 committed:
Highlight text with colors #450

* philippes code is working only for one match

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge branch 'main' of https://github.com/nexB/scancode.io into lf32-licensetext-devel

lf32 commented:
@pombredanne  The `get_matched_text()` is highlighting only one license per text and to render all licenses, I am storing in an array `complete_text_in_array`, so that I can render them all at once. Could you help me fix this? 

https://github.com/nexB/scancode.io/pull/469/commits/cd674fbec747d342fc7c555e8372deff1489d444


![errorex](https://user-images.githubusercontent.com/96695352/181576606-0e2186bd-186a-4afe-910c-ecf21ecda905.png)

lf32 committed:
Fix failing text for `make valid` #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Fix failing test for `make valid` #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge branch 'main' of into lf32-licensetext-devel

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Testing out new UI to match the projects page #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add match_text tests to scantext #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

pombredanne committed:
Make scancode match text test pass

* Some adjustments were needed to ensure we could run these copied tests
  correctly.

* See also this issue that requires to install SCTK locally in SCIO
  using "pip install --editable <path to your scancode checkout"
  and then to run "scancode --reinxdex-licenses" to get proper paths.
  https://github.com/nexB/scancode-toolkit/issues/3044

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add list of matches to Token

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
Some possible simple approaches: 
- https://codepen.io/warkentien2/pen/XKbmBw
- https://alienzhou.github.io/web-highlighter/
- https://github.com/julmot/mark.js
- https://alvarotrigo.com/blog/css-highlight-text/

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Add all detected values into the table #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add details page to view matched license details #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add a mini licenses file to test for development #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Move import of Token to top #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add `license_chart_data` to render charts #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add eye friendly green color to match highlights #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Highlight all detected licenses in one match #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Fix tests by running make valid

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Match token args correction #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add hyperlinks to urls in details #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
What can be added to the details of the license table?

![license_table_1](https://user-images.githubusercontent.com/96695352/185979182-b959b209-9861-42c8-a933-60fb8cdc8401.png)

lf32 commented:
@AyanSinhaMahapatra should the license with matched rule as `unknown-license-detection:7dea8c34d7fd1992015797ce07e2c39d` be hyperlinked to `/rules/` or `/licenses/`?

lf32 committed:
Update details, Improve UI #450

* the details now will use sctk's add-license-detection
branch json styling

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge v31.0.0 of 'main' into lf32-licensetext-devel

lf32 committed:
Ace editor upgraded from v1.4.12 to v1.9.5

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Improve details, highlights and tests

- Rearrange details and improve hyperlinking
- Improve highlighting colors
- Write tests for modules

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
@tdruez I was thinking of using `dropdown_hoverable.html` to show detected licenses on hovering each word. But I end up messing with the layout of input text. Could you help me solve this?

tdruez commented:
@lf32 sure, point me to the code.

lf32 committed:
Add report functionality to the summary

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
Update:
I've added a report feature to the summary page.



![x](https://user-images.githubusercontent.com/96695352/188222115-47fbab08-26b9-43ed-9ea7-bba3c7a076ef.png)

lf32 commented:
> @lf32 sure, point me to the code.

Here it is @tdruez 

https://github.com/nexB/scancode.io/blob/bdae386e83ea8fd5d1a3fda48a29e7867a0f2833/scantext/templates/scantext/includes/license_summary_detail.html#L4

lf32 commented:
@tdruez I'm trying to make the accessibility of license details better and I came up with this idea. 

 Ability to view details on hovering over the arrow down icon instead of clicking (replace bulma css cards with bulma dropdown)

I am doing this only because as I haven't implemented scroll to highlighted line (in colors not ace editor) on dropdown icon click.

![hover](https://user-images.githubusercontent.com/96695352/188422467-47757bdb-3cd7-401f-93aa-452cd0ee6086.png)

tdruez commented:
@lf32 I like the dropdown UI, looks good.

> I am doing this only because as I haven't implemented scroll to highlighted line

We still need a way to see/reach the highlighted text for a selected license.

lf32 committed:
Add new user interface and cleanup the old one

* Ace isn't used any more
* Summary is presented in single page only

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

tdruez commented:
@lf32 could you fixed the failing tests?
See https://github.com/nexB/scancode.io/actions/runs/3008055116

lf32 committed:
Remove unused tests and validate code format #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
@tdruez those tests in `test_match_text.py` are old and need changes(they were taken from toolkit), there are minimal tests for the `scantext` app inside `tests.py`

lf32 commented:
> @lf32 could you fixed the failing tests?

@tdruez should I include those tests?

tdruez commented on scantext/views.py:
```diff
@@ -0,0 +1,395 @@
+# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
+# Visit https://github.com/nexB/scancode.io for support and download.
+
+import sys
+import tempfile
+from pprint import pprint
+
+from django.conf import settings
+from django.contrib import messages
+from django.shortcuts import render
+
+import attr
+from licensedcode import models
+from licensedcode import query
+from licensedcode.spans import Span
+from licensedcode.stopwords import STOPWORDS
+from licensedcode.tokenize import index_tokenizer
+from licensedcode.tokenize import matched_query_text_tokenizer
```
Remove un-used imports, for example:
- import sys
- from pprint import pprint
- from django.conf import settings
- ...

You also have un-used imports within function, such as `from licensedcode.spans import Span`

lf32 replied:
Yes, noticed them. Will remove them.

tdruez commented on scantext/views.py:
```diff
@@ -0,0 +1,395 @@
+
+TRACE_HIGHLIGHTED_TEXT = True
+
+SCANCODE_BASE_URL = (
+    "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data"
+)
+SCANCODE_REPO_URL = "https://github.com/nexB/scancode-toolkit"
+SPDX_LICENSE_URL = "https://spdx.org/licenses/{}"
+DEJACODE_LICENSE_URL = "https://enterprise.dejacode.com/urn/urn:dje:license:{}"
+SCANCODE_LICENSEDB_URL = "https://scancode-licensedb.aboutcode.org/{}"
```
Some of those are not used.

lf32 replied:
Will remove them.

tdruez commented on scantext/views.py:
```diff
@@ -0,0 +1,395 @@
+            },
+        )
+
+    # import json
```
Remove debug code.

lf32 replied:
sure tdruez!

tdruez commented on scantext/views.py:
```diff
@@ -0,0 +1,395 @@
+    )
+
+
+def get_licenses(
```
Should all those non-view functions be moved into the match_text module?

lf32 replied:
Yes tdruez!

tdruez commented on scantext/tests/test_views.py:
```diff
@@ -0,0 +1,76 @@
+        matches = [match1, match2, match3]
+        assert get_license_keys_count(matches) == 4
+
+    def test_get_rule_text_url__for_rule(self):
```
Why the double "_" in the function name?

lf32 replied:
@pombredanne said it would be easy to read.

tdruez replied:
It's not consistent with existing test names in `scanpipe` module nor with Django's conventions https://github.com/django/django/tree/main/tests
Please remove the extra `_`

lf32 replied:
sure tdruez!

tdruez commented on scantext/tests/test_views.py:
```diff
@@ -0,0 +1,76 @@
+        rule1.identifier = "Apache-2.0.RULE"
+        result = get_rule_text_url(rule=rule1, base_url="http://example.com")
+
+        assert result == "http://example.com/rules/Apache-2.0.RULE", result
```
You need to use the same assertion style as already existing in the `scanpipe` module.

tdruez reviewed (changes requested):
Code review:

- 4 `# TODO` still in the code that were not addressed:
  - "# TODO: check this to handle input files"
  - "# TODO: percentage of license text should be done by scancode-toolkit."
  - "# TODO: decide whether the text should be highlighted or not."
  - "# TODO: Is this replacing spdx_key???"
- Templates indentation, not following the convention established in `scanpipe` which is 2 spaces, not 4.
- We are missing unit tests for the following functions:
  - tokenize_matched_text
  - license_scanview
  - get_licenses
  - get_percentage_of_license_text
  - get_license_tokens
  - tag_matched_tokens
  - build_colors
  - get_match_details

tdruez commented:
UI review:

- Everywhere we display the score, we should round up to an integer, the `.0` is not useful
![Screenshot 2022-09-12 at 09 45 21](https://user-images.githubusercontent.com/489057/189581753-d00efc39-2303-4551-8f3d-44b4f7243663.jpg)

- The "Report on GitHub" link is often not working: `Bad request: Your request URL is too long.`. Also, we said previously that we do not want to submit user data to GitHub. 

- The "View more details" modal does not provide much more interesting data than the dropdown. We should improve the dropdown to have links like the modal, and probably get rid of the modal.

- There is not visual connection or clue between the list of "Detected Licenses" on the left and the Highlighted texts on the right.

- The `title` of the highlighted text shows a number instead of license keys:
![Screenshot 2022-09-12 at 09 52 24](https://user-images.githubusercontent.com/489057/189582682-77d774e4-04e1-4db8-9bb8-990eea0cf553.jpg)

lf32 commented:
> The "Report on GitHub" link is often not working: Bad request: Your request URL is too long.. Also, we said previously that we do not want to submit user data to GitHub.

So, Only Detection level details should be reported?

for example, if the license that is to be reported is `apache-2.0.LICENSE` then are the following details enough to report?

```python
{
    score : 100 
    start_line : 1 
    end_line : 201 
    matched_length : 1584 
    match_coverage : 100.0 
    rule_identifier : apache-2.0.LICENSE
}
```

tdruez commented:
> So, Only Detection level details should be reported?

Could you point me to the place were the design for the reporting to GitHub feature was discussed?

lf32 commented on scantext/templates/scantext/includes/license_report.html:
````diff
@@ -0,0 +1,2 @@
+<a class="has-text-danger" href="https://github.com/nexB/scancode.io/issues/new?labels=bug&title=License+detection+error+as+`{{ license.license_expression|pprint }}`
+&body=Detection+level+details%0A```python%0A{%0A%20%20%20%20score+:+{{ license.score }}+%0A%20%20%20%20start_line+:+{{ license.start_line }}+%0A%20%20%20%20end_line+:+{{ license.end_line }}+%0A%20%20%20%20matched_length+:+{{ license.matched_length }}+%0A%20%20%20%20match_coverage+:+{{ license.match_coverage }}+%0A%20%20%20%20rule_identifier+:+{{ license.rule_identifier }}%0A}%0A```+%0A%0AMatched+ ...
````
> Could you point me to the place were the design for the reporting to GitHub feature was discussed?

@tdruez here it is

tdruez replied:
This is not a "design discussion", this is code...

lf32 replied:
Reporting license details should be done at `scancode-toolkit issues` page.

lf32 replied:
The project idea says 
```
It will also allow the integrated reporting of license detection issues in the app based on the results.
```

lf32 committed:
Indent html, handle input, fix modules & tests#450

* Indent html code to 2 spaces
* Create temp directories for scanning
* Move all non-views to `match_text.py`
* Correct existing tests to match the rest of the app
* Remove debug code

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add more licenses, highlight text properly #450

* Added few more licenses to test
* Correct highlighting text
* More styles to highlights

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Set cursor style to help, add token attr desc #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Connect left and right part of ui #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
The ui for the last commit looks like this.

The way it works is
When a panel dropdown button (will change it to the whole block) is clicked or the text (on right) is clicked, the panel block and the matched text both will be highlighted as below. This way the left and right are connected visually.  

![tmp](https://user-images.githubusercontent.com/96695352/191485403-dead9233-69d2-4aac-8237-a04aa63b7330.png)

lf32 committed:
Set predefined colors to 3 #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add more colors & new test licenses

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Add highlight color and get rid of modal #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

tdruez commented:
Hi @lf32 what's the latest status on this PR?

lf32 commented:
@tdruez I am writing the tests.

lf32 commented:
- [ ] write unit tests
- [ ] set a fixed width for the dropdown
- [ ] add navigation with prev and next licenses

tdruez commented:
@lf32 Thanks for the update!

lf32 committed:
Fix failing tests #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Fix width for dropdown #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

DennisClark commented:
per @pombredanne   we need to determine if this properly belongs in ScanCode.io or in a separate application.  The design-to-date is a separate app.

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
Write input file either in chunks or text #450

This feature helps in working with files of type
text or other (example: binary)

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 committed:
Move scan to utility dropdown #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>

lf32 commented:
@AyanSinhaMahapatra, I have added a utilities dropdown and moved `scan` tab to `Detect License` which redirects to `/scantext/`.
![utility](https://user-images.githubusercontent.com/96695352/216820206-7d704256-2509-4325-b1e9-16d24f857b31.png)

tdruez commented:
Hi @lf32 what's the latest status on this PR? Is it feature complete?

lf32 commented:
@tdruez I have discussed with philippe about it and still needs changes to be made.

tdruez commented:
@lf32 Can you provide a list of those changes?

lf32 commented:
- [ ] writing more tests
- [ ] code correction
- [ ] migrating to scancode-toolkit v32 in scantext
- [ ] more efficient user interface

lf32 committed:
Merge branch 'main' into lf32-licensetext-devel

lf32 committed:
(tests) code format is vaildated #450

Signed-off-by: Akhil Raj <lf32.dev@gmail.com>
